File: src/utils/chat/messageFormat.js
import { getFileType } from "@/src/utils/storage/file/type";
import useUserStore from "@/src/context/UserContext";
import i18n from "@/src/i18n";
import { DateTime } from "luxon";

const format = (messageRef) => {
  if (!messageRef) return messageRef;

  const message = { ...messageRef };

  if (message && message.type) {
    if (message.type == "system") {
      message.content = getSystemMessageText(message);
    } else if (message.type == "message" || message.type == "DRAFT") {
      if (!message.content || message.content.trim() === "") {
        if (message.files && message.files.length > 0) {
          const types = message.files.map((file) =>
            getFileType(file.mimeType || file.type, file.name),
          );
          const uniqueTypes = [...new Set(types)];
          if (uniqueTypes.length === 1) {
            const type = uniqueTypes[0];
            const count = types.length;
            const fileTypeMap = {
              IMAGE: {
                emoji: "📷",
                singular: i18n.t("messageFormat.fileType.image.singular"),
                plural: i18n.t("messageFormat.fileType.image.plural"),
              },
              VIDEO: {
                emoji: "📹",
                singular: i18n.t("messageFormat.fileType.video.singular"),
                plural: i18n.t("messageFormat.fileType.video.plural"),
              },
              AUDIO: {
                emoji: "🎵",
                singular: i18n.t("messageFormat.fileType.audio.singular"),
                plural: i18n.t("messageFormat.fileType.audio.plural"),
              },
              VOICE: {
                emoji: "🎤",
                singular: i18n.t("messageFormat.fileType.voice.singular"),
                plural: i18n.t("messageFormat.fileType.voice.plural"),
              },
              DOCUMENT: {
                emoji: "📄",
                singular: i18n.t("messageFormat.fileType.document.singular"),
                plural: i18n.t("messageFormat.fileType.document.plural"),
              },
              CODE: {
                emoji: "💻",
                singular: i18n.t("messageFormat.fileType.code.singular"),
                plural: i18n.t("messageFormat.fileType.code.plural"),
              },
              ARCHIVE: {
                emoji: "🗄️",
                singular: i18n.t("messageFormat.fileType.archive.singular"),
                plural: i18n.t("messageFormat.fileType.archive.plural"),
              },
            };
            const { emoji, singular, plural } = fileTypeMap[type] || {
              emoji: "📎",
              singular: i18n.t("messageFormat.fileType.default.singular"),
              plural: i18n.t("messageFormat.fileType.default.plural"),
            };

            let durationStr = "";
            if (
              count === 1 &&
              (type === "AUDIO" || type === "VIDEO" || type === "VOICE")
            ) {
              const file = message.files[0];
              if (file && file.duration) {
                const seconds = file.duration;
                const m = Math.floor(seconds / 60);
                const s = Math.floor(seconds % 60);
                durationStr = ` ${m}:${s.toString().padStart(2, "0")}`;
              }
            }

            message.content =
              count === 1
                ? `${emoji} ${singular}${durationStr}`
                : `${count} ${emoji} ${plural}`;
          } else {
            const hasOnlyMedia = uniqueTypes.every(
              (type) => type === "IMAGE" || type === "VIDEO",
            );
            message.content = hasOnlyMedia
              ? `${message.files.length} 📎 ${i18n.t("messageFormat.media")}`
              : `${message.files.length} 📎 ${i18n.t("messageFormat.files")}`;
          }
        }
      } else {
        const gifUrls = extractGifUrls(message.content);
        const textWithoutGifs = stripGifUrls(message.content);
        if (gifUrls.length > 0 && !textWithoutGifs) {
          message.content =
            gifUrls.length === 1
              ? `🎞️ ${i18n.t("messageFormat.fileType.gif.singular")}`
              : `${gifUrls.length} 🎞️ ${i18n.t("messageFormat.fileType.gif.plural")}`;
        }
      }
    }
  }
  return message;
};

/**
 * Get real text from system message object
 * @param {Object} message
 * @returns {String} text
 */
const getSystemMessageText = (message) => {
  let text = "";
  let name = "User";

  const { localUserUUID, getUser } = useUserStore.getState();

  switch (message.system_action) {
    case "CHAT_CREATED":
      text = i18n.t("messageFormat.system.chatCreated");
      break;
    case "USER_JOINED":
      if (message.content === localUserUUID) {
        name = i18n.t("messageFormat.system.you");
      } else {
        const user = getUser(message.content);
        name = user ? user.name : i18n.t("messageFormat.system.user");
      }
      text = i18n.t("messageFormat.system.userJoined", { name });
      break;
    case "USER_LEFT":
      if (message.content === localUserUUID) {
        name = i18n.t("messageFormat.system.you");
      } else {
        const user = getUser(message.content);
        name = user ? user.name : i18n.t("messageFormat.system.user");
      }
      text = i18n.t("messageFormat.system.userLeft", { name });
      break;
    default:
      text = i18n.t("messageFormat.system.systemMessage");
  }
  return text;
};

/**
 * Get real text from member activity data
 * @param {Array} memberActivityData
 * @returns {String} text
 */
const formatActivity = (memberActivityData, chatType) => {
  if (!memberActivityData || memberActivityData.length === 0) return "";

  const { localUserUUID, getUser } = useUserStore.getState();

  const activeActivities = memberActivityData.filter(
    (a) => a.action && a.userUUID !== localUserUUID,
  );

  if (activeActivities.length === 0) return "";

  // Group by action to find majority
  const actionsMap = {};
  const actionOrder = [];
  activeActivities.forEach((a) => {
    if (!actionsMap[a.action]) {
      actionsMap[a.action] = [];
      actionOrder.push(a.action);
    }
    actionsMap[a.action].push(a);
  });

  // Find the action with the most participants (majority)
  let majorityAction = actionOrder[0];
  let maxCount = actionsMap[majorityAction].length;

  actionOrder.forEach((action) => {
    const count = actionsMap[action].length;
    if (count > maxCount) {
      maxCount = count;
      majorityAction = action;
    }
  });

  const participants = actionsMap[majorityAction];
  const count = participants.length;
  const names = participants.map((p) => {
    const user = getUser(p.userUUID);
    return user ? user.name : i18n.t("messageFormat.system.user");
  });

  const getActionI18nKey = (action) => {
    switch (action) {
      case "TYPING":
        return "typing";
      case "RECORDING_VOICE":
        return "recording_voice";
      case "RECORDING_VIDEO":
        return "recording_video";
      case "UPLOADING_FILE":
        return "uploading_file";
      default:
        return "active";
    }
  };

  const actionKey = getActionI18nKey(majorityAction);

  let formattedActivity = "";
  if (count === 1) {
    formattedActivity = i18n.t(`messageFormat.activity.${actionKey}_one`, {
      name: names[0],
    });
  } else if (count === 2) {
    formattedActivity = i18n.t(`messageFormat.activity.${actionKey}_two`, {
      name: names[0],
      name2: names[1],
    });
  } else {
    formattedActivity = i18n.t(`messageFormat.activity.${actionKey}_other`, {
      name: names[0],
      name2: names[1],
      count: count - 2,
    });
  }

  return formattedActivity;
};

/**
 * Format a date for "last seen" status
 * @param {Date|String} lastAccessAt
 * @returns {String} formatted text
 */
const formatLastSeen = (lastAccessAt) => {
  if (!lastAccessAt) return "";

  try {
    const date = DateTime.fromISO(new Date(lastAccessAt).toISOString(), {
      zone: "utc",
    }).toLocal();
    const now = DateTime.now();
    const diffMs = now.toMillis() - date.toMillis();

    const diffSecs = Math.floor(diffMs / 1000);
    const diffMins = Math.floor(diffSecs / 60);
    const diffHours = Math.floor(diffMins / 60);
    const diffDays = Math.floor(diffHours / 24);

    if (diffSecs < 10) return i18n.t("messageFormat.time.justNow");
    if (diffSecs < 60)
      return i18n.t("messageFormat.time.secondsAgo", { count: diffSecs });
    if (diffMins < 60)
      return i18n.t("messageFormat.time.minutesAgo", { count: diffMins });
    if (diffHours < 24)
      return i18n.t("messageFormat.time.hoursAgo", { count: diffHours });
    if (diffDays === 1) return i18n.t("messageFormat.time.yesterday");
    if (diffDays < 7)
      return i18n.t("messageFormat.time.daysAgo", { count: diffDays });

    return date.setLocale(i18n.language).toLocaleString(DateTime.DATE_FULL);
  } catch (e) {
    return "";
  }
};

/** .gif link */
const GIF_URL_REGEX = /https?:\/\/[^\s<>"'`]+?\.gif(?:\?[^\s<>"'`]*)?/gi;

const normalizeGifUrl = (raw) =>
  raw
    .trim()
    .replace(/[),.;!?]+$/, "")
    .replace(/^http:/, "https:");

const extractGifUrls = (content) => {
  if (!content) return [];
  GIF_URL_REGEX.lastIndex = 0;
  const matches = content.match(GIF_URL_REGEX) || [];
  const seen = new Set();
  const result = [];
  for (const raw of matches) {
    const url = normalizeGifUrl(raw);
    if (!seen.has(url)) {
      seen.add(url);
      result.push(url);
    }
  }
  return result;
};

const stripGifUrls = (content) => {
  if (!content) return "";
  return content
    .replace(GIF_URL_REGEX, " ")
    .replace(/[ \t]+\n/g, "\n")
    .replace(/\n{3,}/g, "\n\n")
    .replace(/[ \t]{2,}/g, " ")
    .trim();
};

const getGifMediaUrl = (url) => {
  if (!url || typeof url !== "string") return null;
  const normalized = normalizeGifUrl(url);
  if (/\.gif(?:\?|$)/i.test(normalized)) return normalized;
  return null;
};

export default {
  format,
  getSystemMessageText,
  formatActivity,
  formatLastSeen,
  extractGifUrls,
  stripGifUrls,
  getGifMediaUrl,
};

File: lib/core/chat/message_format.dart
import 'package:intl/intl.dart';
import 'package:novyse/core/l10n/l10n.dart';
import 'package:novyse/core/storage/file/file_type.dart';
import 'package:novyse/core/stores/message_store.dart';

final RegExp _gifUrlRegex = RegExp(
  r'''https?://[^\s<>"'`]+?\.gif(?:\?[^\s<>"'`]*)?''',
  caseSensitive: false,
);

String _normalizeGifUrl(String raw) {
  return raw
      .trim()
      .replaceAll(RegExp(r'[),.;!?]+$'), '')
      .replaceFirst(RegExp(r'^http:'), 'https:');
}

/// Extracts unique GIF URLs from string content.
List<String> extractGifUrls(String? content) {
  if (content == null || content.isEmpty) return [];
  final matches = _gifUrlRegex.allMatches(content);
  final seen = <String>{};
  final result = <String>[];

  for (final match in matches) {
    final raw = match.group(0);
    if (raw != null) {
      final url = _normalizeGifUrl(raw);
      if (seen.add(url)) {
        result.add(url);
      }
    }
  }
  return result;
}

/// Strips all GIF URLs from the message text.
String stripGifUrls(String? content) {
  if (content == null || content.isEmpty) return '';
  return content
      .replaceAll(_gifUrlRegex, ' ')
      .replaceAll(RegExp(r'[ \t]+\n'), '\n')
      .replaceAll(RegExp(r'\n{3,}'), '\n\n')
      .replaceAll(RegExp(r'[ \t]{2,}'), ' ')
      .trim();
}

/// Returns the normalized GIF URL if the string points to a valid GIF.
String? getGifMediaUrl(String? url) {
  if (url == null || url.isEmpty) return null;
  final normalized = _normalizeGifUrl(url);
  if (RegExp(r'\.gif(?:\?|$)', caseSensitive: false).hasMatch(normalized)) {
    return normalized;
  }
  return null;
}

/// Formats message text / preview content (e.g. attachments, GIFs, system messages).
Map<String, dynamic> formatMessage(
  dynamic messageRef, {
  required AppLocalizations l10n,
  String? localUserUUID,
  Map<String, dynamic>? Function(String uuid)? getUser,
}) {
  final Map<String, dynamic> message = messageRef is MessageModel
      ? messageRef.toMap()
      : (messageRef is Map
            ? Map<String, dynamic>.from(messageRef)
            : <String, dynamic>{});
  final type = message['type'] as String?;

  if (type == 'system') {
    message['content'] = getSystemMessageText(
      message,
      l10n: l10n,
      localUserUUID: localUserUUID,
      getUser: getUser,
    );
  } else if (type == 'message' || type == 'DRAFT') {
    final content = message['content'] as String?;
    if (content == null || content.trim().isEmpty) {
      final files = message['files'];
      if (files is List && files.isNotEmpty) {
        final fileCategories = <FileTypeCategory>[];
        for (final f in files) {
          if (f is Map) {
            final mime = (f['mimeType'] ?? f['type'] ?? '') as String;
            final name = (f['name'] ?? f['fileName'] ?? '') as String;
            fileCategories.add(getFileType(mime, name));
          }
        }

        final uniqueCategories = fileCategories.toSet().toList();
        if (uniqueCategories.length == 1) {
          final cat = uniqueCategories.first;
          final count = fileCategories.length;

          final (emoji, singular, plural) = switch (cat) {
            FileTypeCategory.image => (
              '📷',
              l10n.msgFormatPhotoSingular,
              l10n.msgFormatPhotoPlural,
            ),
            FileTypeCategory.video => (
              '📹',
              l10n.msgFormatVideoSingular,
              l10n.msgFormatVideoPlural,
            ),
            FileTypeCategory.audio => (
              '🎵',
              l10n.msgFormatAudioSingular,
              l10n.msgFormatAudioPlural,
            ),
            FileTypeCategory.voice => (
              '🎤',
              l10n.msgFormatVoiceSingular,
              l10n.msgFormatVoicePlural,
            ),
            FileTypeCategory.document => (
              '📄',
              l10n.msgFormatDocumentSingular,
              l10n.msgFormatDocumentPlural,
            ),
            FileTypeCategory.code => (
              '💻',
              l10n.msgFormatCodeSingular,
              l10n.msgFormatCodePlural,
            ),
            FileTypeCategory.archive => (
              '🗄️',
              l10n.msgFormatArchiveSingular,
              l10n.msgFormatArchivePlural,
            ),
            FileTypeCategory.other => (
              '📎',
              l10n.msgFormatFileSingular,
              l10n.msgFormatFilePlural,
            ),
          };

          var durationStr = '';
          if (count == 1 &&
              (cat == FileTypeCategory.audio ||
                  cat == FileTypeCategory.video ||
                  cat == FileTypeCategory.voice)) {
            final file = files.first;
            if (file is Map && file['duration'] != null) {
              final seconds = (file['duration'] as num).toInt();
              final m = seconds ~/ 60;
              final s = seconds % 60;
              durationStr = ' $m:${s.toString().padLeft(2, '0')}';
            }
          }

          message['content'] = count == 1
              ? '$emoji $singular$durationStr'
              : '$count $emoji $plural';
        } else {
          final hasOnlyMedia = uniqueCategories.every(
            (c) => c == FileTypeCategory.image || c == FileTypeCategory.video,
          );
          message['content'] = hasOnlyMedia
              ? '${files.length} 📎 ${l10n.msgFormatMedia}'
              : '${files.length} 📎 ${l10n.msgFormatFiles}';
        }
      }
    } else {
      final gifUrls = extractGifUrls(content);
      final textWithoutGifs = stripGifUrls(content);
      if (gifUrls.isNotEmpty && textWithoutGifs.isEmpty) {
        message['content'] = gifUrls.length == 1
            ? '🎞️ ${l10n.msgFormatGifSingular}'
            : '${gifUrls.length} 🎞️ ${l10n.msgFormatGifPlural}';
      }
    }
  }

  return message;
}

/// Formats readable system action text for system messages.
String getSystemMessageText(
  Map<String, dynamic> message, {
  required AppLocalizations l10n,
  String? localUserUUID,
  Map<String, dynamic>? Function(String uuid)? getUser,
}) {
  final action = message['system_action'] as String?;
  final content = (message['content'] ?? '') as String;

  String resolveName() {
    if (content == localUserUUID) return l10n.chatYou;
    final user = getUser?.call(content);
    return (user?['name'] as String?) ?? l10n.user;
  }

  switch (action) {
    case 'CHAT_CREATED':
      return l10n.msgFormatChatCreated;
    case 'USER_JOINED':
      return l10n.msgFormatUserJoined(resolveName());
    case 'USER_LEFT':
      return l10n.msgFormatUserLeft(resolveName());
    default:
      return l10n.msgFormatSystemMessage;
  }
}

/// Formats typing and member activity data.
String formatActivity(
  List<dynamic> memberActivityData, {
  required AppLocalizations l10n,
  String? localUserUUID,
  Map<String, dynamic>? Function(String uuid)? getUser,
}) {
  if (memberActivityData.isEmpty) return '';

  final activeActivities = memberActivityData.where((a) {
    if (a is! Map) return false;
    final action = a['action'];
    final userUUID = a['userUUID'];
    return action != null && userUUID != localUserUUID;
  }).toList();

  if (activeActivities.isEmpty) return '';

  final actionsMap = <String, List<Map<String, dynamic>>>{};
  final actionOrder = <String>[];

  for (final a in activeActivities) {
    final item = Map<String, dynamic>.from(a as Map);
    final action = item['action'] as String;
    if (!actionsMap.containsKey(action)) {
      actionsMap[action] = [];
      actionOrder.add(action);
    }
    actionsMap[action]!.add(item);
  }

  var majorityAction = actionOrder.first;
  var maxCount = actionsMap[majorityAction]!.length;

  for (final action in actionOrder) {
    final count = actionsMap[action]!.length;
    if (count > maxCount) {
      maxCount = count;
      majorityAction = action;
    }
  }

  final participants = actionsMap[majorityAction]!;
  final count = participants.length;
  final names = participants.map((p) {
    final uuid = (p['userUUID'] ?? '') as String;
    final user = getUser?.call(uuid);
    return (user?['name'] as String?) ?? l10n.user;
  }).toList();

  final name = names[0];
  final name2 = count > 1 ? names[1] : '';
  final others = count - 2;

  switch (majorityAction) {
    case 'TYPING':
      if (count == 1) return l10n.msgFormatTypingOne(name);
      if (count == 2) return l10n.msgFormatTypingTwo(name, name2);
      return l10n.msgFormatTypingOther(name, name2, others);
    case 'RECORDING_VOICE':
      if (count == 1) return l10n.msgFormatRecordingVoiceOne(name);
      if (count == 2) return l10n.msgFormatRecordingVoiceTwo(name, name2);
      return l10n.msgFormatRecordingVoiceOther(name, name2, others);
    case 'RECORDING_VIDEO':
      if (count == 1) return l10n.msgFormatRecordingVideoOne(name);
      if (count == 2) return l10n.msgFormatRecordingVideoTwo(name, name2);
      return l10n.msgFormatRecordingVideoOther(name, name2, others);
    case 'UPLOADING_FILE':
      if (count == 1) return l10n.msgFormatUploadingFileOne(name);
      if (count == 2) return l10n.msgFormatUploadingFileTwo(name, name2);
      return l10n.msgFormatUploadingFileOther(name, name2, others);
    default:
      if (count == 1) return l10n.msgFormatActiveOne(name);
      if (count == 2) return l10n.msgFormatActiveTwo(name, name2);
      return l10n.msgFormatActiveOther(name, name2, others);
  }
}

/// Formats relative time for user "last seen" status.
String formatLastSeen(
  dynamic lastAccessAt, {
  required AppLocalizations l10n,
}) {
  if (lastAccessAt == null) return '';

  try {
    DateTime date;
    if (lastAccessAt is DateTime) {
      date = lastAccessAt.toLocal();
    } else if (lastAccessAt is String) {
      date = DateTime.parse(lastAccessAt).toLocal();
    } else if (lastAccessAt is int) {
      date = DateTime.fromMillisecondsSinceEpoch(lastAccessAt).toLocal();
    } else {
      return '';
    }

    final now = DateTime.now();
    final diff = now.difference(date);

    if (diff.inSeconds < 10) {
      return l10n.msgFormatJustNow;
    }
    if (diff.inSeconds < 60) {
      return l10n.msgFormatSecondsAgo(diff.inSeconds);
    }
    if (diff.inMinutes < 60) {
      return l10n.msgFormatMinutesAgo(diff.inMinutes);
    }
    if (diff.inHours < 24) {
      return l10n.msgFormatHoursAgo(diff.inHours);
    }
    if (diff.inDays == 1) {
      return l10n.msgFormatYesterday;
    }
    if (diff.inDays < 7) {
      return l10n.msgFormatDaysAgo(diff.inDays);
    }

    return DateFormat.yMd(l10n.localeName).format(date);
  } catch (_) {
    return '';
  }
}
